bitcoin/src/bip143.rs

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: CC0-1.0
+// Written by the rust-bitcoin developers.
+
+//! [BIP143] Transaction Signature Verification for Version 0 Witness Program
+//!
+//! [BIP143]: <https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki>
+//!
+
+use core::fmt;
+
+use crate::blockdata::opcodes::all::*;
+use crate::consensus::{encode, Decodable, Encodable};
+use crate::consensus::encode::WriteExt;
+use crate::io;
+use crate::prelude::*;
+use crate::script::{Builder, Script, ScriptBuf};
+
+/// The scriptCode as defined in [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki).
+#[derive(Default, Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct ScriptCode(ScriptBuf);
+
+impl ScriptCode {
+    /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
+    /// for a P2WPKH output.
+    pub fn new_p2wpkh(wpkh: &Script) -> Option<Self> {
```

Comment:
This should not be `&Script`. It should either be `WPubKeyHash` or preferably `PublicKey`.

Reply:
Just saw this comment @sanket1729, posting so you know I haven't ignored it. I had a quick go but I'm too brain fried to add it today. I'll make a note though. Thanks